packages/frontend/core/src/utils/navigable-history.ts

```diff
@@ -0,0 +1,221 @@
+import type { Blocker, Listener, Location, To } from 'history';
+import {
+  Action,
+  createPath,
+  type MemoryHistory,
+  type MemoryHistoryOptions,
+  parsePath,
+} from 'history';
+
+export interface NavigableHistory extends MemoryHistory {
```

Comment:
what's the difference from react-router's memory history

Reply:
react-router v6 only has MemoryRouter, which internally uses MemoryHistory of 'history' package.

Reply:
MemoryRouter also includes some async loader and other functions, making it very complex and difficult to use.